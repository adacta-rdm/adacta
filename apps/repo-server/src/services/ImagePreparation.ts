import assert from "assert";

import { assertUnreachable } from "@omegadot/assert";
import type { FormatEnum } from "sharp";

import type { IImageOptions } from "~/apps/repo-server/interface/IPrepareImageTaskArgs";
import { IImagePreset } from "~/apps/repo-server/src/graphql/generated/resolvers";
import { TaskDispatcher } from "~/apps/repo-server/src/services/TaskDispatcher/TaskDispatcher";
import type { StorageEngineRemoteAccess } from "~/apps/repo-server/src/storage/storageEngine/remoteAccess/StorageEngineRemoteAccess";
import type { DrizzleEntity } from "~/drizzle/DrizzleSchema";
import type { IResourceId } from "~/lib/database/Ids";
import { Service, ServiceContainer } from "~/lib/serviceContainer/ServiceContainer";
import type { StorageEngine } from "~/lib/storage-engine";

/**
 * This service is responsible for preparing images for display.
 * It will take an image resource and does the following:
 * 	- Resize the image to the desired size
 * 	- Convert the image to suitable format
 * 	- If necessary the image is rotated to the correct orientation (provided by the EXIF data)
 */
@Service()
export class ImagePreparation {
	private map = new Set<string>();

	/**
	 * Filetypes supported by sharp
	 * @private
	 */
	private static sharpExtension: Partial<Record<keyof FormatEnum, string>> = {
		avif: "image/avif",
		gif: "image/gif",
		jpeg: "image/jpeg",
		jpg: "image/jpeg",
		pdf: "application/pdf",
		png: "image/png",
		svg: "image/svg+xml",
		tiff: "image/tiff",
		tif: "image/tiff",
		webp: "image/webp",
	};

	/**
	 * Filetypes with custom handling
	 * @private
	 */
	private static customExtensions = {
		heic: "image/heic",
		heif: "image/heif",
		bmp: "image/bmp",
	};

	private static supportedMimeTypes: string[] = Object.values({
		...ImagePreparation.sharpExtension,
		...ImagePreparation.customExtensions,
	});

	private static IMAGE_SIZES = {
		icon: { width: 25, height: 25 },
		thumbnail: { width: 200, height: 200 },
		regular: { width: 1920, height: 1080 },
	};

	public async getImage(
		resource: DrizzleEntity<"Resource">,
		sto: StorageEngine,
		stoRemote: StorageEngineRemoteAccess,
		preset: IImagePreset
	) {
		return {
			url: await this.requestImage(resource, sto, stoRemote, preset),
			fallbackURL: await this.requestImage(resource, sto, stoRemote, preset),
		};
	}

	private static isPictureType(type: string) {
		const t = type.toLowerCase();
		return t === "jpg" || t === "jpeg" || t === "heic";
	}

	public static isSupportedMimeType(mime: string) {
		return ImagePreparation.supportedMimeTypes.includes(mime);
	}

	private static getOptions(preset: IImagePreset, extension: string): IImageOptions {
		switch (preset) {
			case IImagePreset.Icon:
				return { type: "png", maxDimensions: ImagePreparation.IMAGE_SIZES["icon"] };
			case IImagePreset.Thumbnail:
				if (ImagePreparation.isPictureType(extension))
					return { type: "jpg", maxDimensions: ImagePreparation.IMAGE_SIZES["thumbnail"] };
				return { type: "png", maxDimensions: ImagePreparation.IMAGE_SIZES["thumbnail"] };
			case IImagePreset.Regular:
				if (ImagePreparation.isPictureType(extension))
					return { type: "jpg", maxDimensions: ImagePreparation.IMAGE_SIZES["regular"] };
				return { type: "png", maxDimensions: ImagePreparation.IMAGE_SIZES["regular"] };
			default:
				assertUnreachable(preset);
		}
	}

	public async requestImage(
		resource: DrizzleEntity<"Resource">,
		sto: StorageEngine,
		stoRemote: StorageEngineRemoteAccess,
		preset: IImagePreset
	) {
		assert(resource.attachment.type === "Image");
		const options = ImagePreparation.getOptions(preset, resource.attachment.mimeType.split("/")[1]);
		const imagePath = ImagePreparation.getPath(resource.id, options);

		const getImageLink = (path: string) => {
			return stoRemote.getDownloadLink(path, undefined, 60 * 60 * 60 * 24 * 7);
		};

		// If the image is already prepared, return the download link directly without a call
		// to S3
		if (this.map.has(imagePath)) {
			return getImageLink(imagePath);
		}

		// If the image is available in the storage engine, return the download link directly
		// without a call to
		if (await sto.exists(imagePath)) {
			this.map.add(imagePath);
			return getImageLink(imagePath);
		}

		const downloadLink = await stoRemote.getDownloadLink(resource.id);
		const uploadLink = await stoRemote.getUploadLink(imagePath);

		await ServiceContainer.get(TaskDispatcher).dispatch("images/prepare", {
			input: {
				inputDownloadURL: downloadLink,
				resultUploadURL: uploadLink,
				originalMimeType: resource.attachment.mimeType,
				options,
			},
		});

		this.map.add(imagePath);

		return getImageLink(imagePath);
	}

	private static getPath(resourceId: IResourceId, options: IImageOptions) {
		const sizeString = `${options.maxDimensions.height}x${options.maxDimensions.width}`;
		const typeString = options.type;

		return `images/${resourceId}_${sizeString}.${typeString}`;
	}
}
