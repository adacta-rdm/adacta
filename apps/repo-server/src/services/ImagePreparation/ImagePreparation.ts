import assert from "assert";

import { assertUnreachable } from "@omegadot/assert";
import type { FormatEnum } from "sharp";

import type { IImageOptions } from "~/apps/repo-server/interface/IPrepareImageTaskArgs";
import { IImagePreset } from "~/apps/repo-server/src/graphql/generated/resolvers";
import { TaskDispatcher } from "~/apps/repo-server/src/services/TaskDispatcher/TaskDispatcher";
import type { StorageEngineRemoteAccess } from "~/apps/repo-server/src/storage/storageEngine/remoteAccess/StorageEngineRemoteAccess";
import type { DrizzleEntity } from "~/drizzle/DrizzleSchema";
import type { IResourceId } from "~/lib/database/Ids";
import { Service } from "~/lib/serviceContainer/ServiceContainer";
import type { StorageEngine } from "~/lib/storage-engine";

/**
 * This service is responsible for preparing images for display.
 * It will take an image resource and does the following:
 *    - Resize the image to the desired size
 *    - Convert the image to suitable format (JPG or PNG, depending on the source image and the
 *    	desired size)
 *    - If necessary the image is rotated to the correct orientation (provided by the EXIF data)
 *
 * The actual image processing is delegated to the `images/prepare` microservice using the
 * TaskDispatcher.
 */
@Service(TaskDispatcher)
export class ImagePreparation {
	/**
	 * Filetypes supported by sharp
	 * @private
	 */
	private static sharpExtension: Partial<Record<keyof FormatEnum, string>> = {
		avif: "image/avif",
		gif: "image/gif",
		jpeg: "image/jpeg",
		jpg: "image/jpeg",
		pdf: "application/pdf",
		png: "image/png",
		svg: "image/svg+xml",
		tiff: "image/tiff",
		tif: "image/tiff",
		webp: "image/webp",
	};
	/**
	 * Filetypes with custom handling
	 * @private
	 */
	private static customExtensions = {
		heic: "image/heic",
		heif: "image/heif",
		bmp: "image/bmp",
	};
	private static supportedMimeTypes: string[] = Object.values({
		...ImagePreparation.sharpExtension,
		...ImagePreparation.customExtensions,
	});
	private static IMAGE_SIZES = {
		icon: { width: 25, height: 25 },
		thumbnail: { width: 200, height: 200 },
		regular: { width: 1920, height: 1080 },
	};
	private map = new Set<string>();

	constructor(private td: TaskDispatcher) {}

	public static isSupportedMimeType(mime: string) {
		return ImagePreparation.supportedMimeTypes.includes(mime);
	}

	private static isPictureType(type: string) {
		const t = type.toLowerCase();
		return t === "jpg" || t === "jpeg" || t === "heic";
	}

	private static getOptions(preset: IImagePreset, extension: string): IImageOptions {
		switch (preset) {
			case IImagePreset.Icon:
				return { type: "png", maxDimensions: ImagePreparation.IMAGE_SIZES["icon"] };
			case IImagePreset.Thumbnail:
				if (ImagePreparation.isPictureType(extension))
					return { type: "jpg", maxDimensions: ImagePreparation.IMAGE_SIZES["thumbnail"] };
				return { type: "png", maxDimensions: ImagePreparation.IMAGE_SIZES["thumbnail"] };
			case IImagePreset.Regular:
				if (ImagePreparation.isPictureType(extension))
					return { type: "jpg", maxDimensions: ImagePreparation.IMAGE_SIZES["regular"] };
				return { type: "png", maxDimensions: ImagePreparation.IMAGE_SIZES["regular"] };
			default:
				assertUnreachable(preset);
		}
	}

	private static getPath(resourceId: IResourceId, options: IImageOptions) {
		const sizeString = `${options.maxDimensions.height}x${options.maxDimensions.width}`;
		const typeString = options.type;

		return `images/${resourceId}_${sizeString}.${typeString}`;
	}

	/**
	 * Prepare an image for display (@see {@link ImagePreparation})
	 * @param resource The image resource
	 * @param preset The preset that describes the desired image size
	 * @param sto
	 * @param stoRemote
	 * @returns The link to the prepared image
	 */
	public async getImage(
		resource: DrizzleEntity<"Resource">,
		preset: IImagePreset,
		sto: StorageEngine,
		stoRemote: StorageEngineRemoteAccess
	) {
		assert(resource.attachment.type === "Image");
		const options = ImagePreparation.getOptions(preset, resource.attachment.mimeType.split("/")[1]);
		const imagePath = ImagePreparation.getPath(resource.id, options);

		const getImageLink = (path: string) => {
			return stoRemote.getDownloadLink(path, undefined, 60 * 60 * 60 * 24 * 7);
		};

		// If the image is already prepared, return the download link directly without a call
		// to S3
		if (this.map.has(imagePath)) {
			return getImageLink(imagePath);
		}

		// If the image is available in the storage engine, return the download link directly
		// without a call to
		if (await sto.exists(imagePath)) {
			this.map.add(imagePath);
			return getImageLink(imagePath);
		}

		const downloadLink = await stoRemote.getDownloadLink(resource.id);
		const uploadLink = await stoRemote.getUploadLink(imagePath);

		await this.td.dispatch("images/prepare", {
			input: {
				inputDownloadURL: downloadLink,
				resultUploadURL: uploadLink,
				originalMimeType: resource.attachment.mimeType,
				options,
			},
		});

		this.map.add(imagePath);

		return getImageLink(imagePath);
	}
}
