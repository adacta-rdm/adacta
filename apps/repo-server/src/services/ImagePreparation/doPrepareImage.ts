import bmpDecode from "@vingle/bmp-js";
import heicDecode from "heic-decode";
import sharp from "sharp";

import type { IImageOptions } from "~/apps/repo-server/interface/IPrepareImageTaskArgs";
import type { Logger } from "~/lib/logger/Logger";

/**
 * Handles the actual image processing for the image preparation service.
 * @param inputDownloadURL url to access the image
 * @param inputMimeType mime type of the input image
 * @param options additional options for image processing (e.g. type, max dimensions)
 * @param logger
 */
export async function doPrepareImage(
	inputDownloadURL: string,
	inputMimeType: string,
	options: IImageOptions,
	logger: Logger
) {
	const response = await fetch(inputDownloadURL);
	const data = await response.arrayBuffer();

	let sharpInstance!: sharp.Sharp;

	logger.debug(`Processing image with type: '${inputMimeType}'`);

	const mimeType = inputMimeType.toLowerCase();

	if (mimeType.includes("heif") || mimeType.includes("heic")) {
		// Sharp does not support HEIC images, when the prebuilt "libvips" is used.
		// To handle HEIC images (and HEIF images as well), we use the "heic-decode" package.
		const buffer = Buffer.from(new Uint8Array(data));
		const decodeResult = await heicDecode({ buffer });
		const { data: rawData, width, height } = decodeResult;
		sharpInstance = sharp(rawData, { raw: { width, height, channels: 4 } });
	} else if (mimeType.includes("bmp")) {
		// Decode BMP images using the "bmp-js" package
		const buffer = Buffer.from(new Uint8Array(data));
		const bitmap = bmpDecode.decode(buffer, true);
		sharpInstance = sharp(bitmap.data, {
			raw: {
				width: bitmap.width,
				height: bitmap.height,
				channels: 4,
			},
		});
	} else {
		sharpInstance = sharp(data, {
			// failOn controls when to abort processing of invalid pixel data
			// This is set to "none" to process the image "as is" without throwing an error
			failOn: "none",
		});
	}

	const selectType = (sharp: sharp.Sharp, type: "jpg" | "png" | "webp") => {
		switch (type) {
			case "jpg":
				return sharp.jpeg();
			case "png":
				return sharp.png();
			case "webp":
				return sharp.webp();
		}
	};

	const transformer = selectType(
		sharpInstance
			.resize({
				width: options.maxDimensions.width,
				height: options.maxDimensions.height,
				fit: sharp.fit.inside, // Keep aspect ratio with width and height as maximum values
			})
			.rotate(), // Auto-rotate based on EXIF data
		options.type
	);
	return transformer.toBuffer();
}
