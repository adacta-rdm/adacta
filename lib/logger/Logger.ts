import type { Writable } from "stream";

import type { JsonObject } from "type-fest";

export enum LOG_LEVEL {
	SILENT,
	FATAL,
	ERROR,
	WARN,
	INFO,
	DEBUG,
	TRACE,
}

interface ILoggerOptions {
	level?: LOG_LEVEL;
	stream: Writable;
}

/**
 * A simple logger interface that writes to the specified stream.
 */
export class Logger {
	private level: LOG_LEVEL;
	private stream: Writable;

	constructor(options: ILoggerOptions, private mergingObject: JsonObject = {}) {
		this.level = options.level ?? LOG_LEVEL.INFO;
		this.stream = options.stream;
	}

	bind(mergingObject: JsonObject): Logger {
		const bound = Object.create(Logger.prototype) as Logger;
		Object.assign(bound, this, { mergingObject: { ...this.mergingObject, ...mergingObject } });
		return bound;
	}

	trace(msg: string): void {
		this.log(LOG_LEVEL.TRACE, msg);
	}

	debug(msg: string): void {
		this.log(LOG_LEVEL.DEBUG, msg);
	}

	info(msg: string): void {
		this.log(LOG_LEVEL.INFO, msg);
	}

	error(msg: string): void {
		this.log(LOG_LEVEL.ERROR, msg);
	}

	warn(msg: string): void {
		this.log(LOG_LEVEL.WARN, msg);
	}

	protected log(level: LOG_LEVEL, msg: string): void {
		if (level <= this.level) {
			this.stream.write(
				`${JSON.stringify({ level, time: Date.now(), ...this.mergingObject, msg })}\n`
			);
		}
	}
}
