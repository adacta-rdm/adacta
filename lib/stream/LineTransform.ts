import type { TransformCallback } from "stream";
import { Transform } from "stream";

/**
 *
 */
export class LineTransform extends Transform {
	private lineCharacters: string[] = [];

	constructor() {
		super({ readableObjectMode: true });
	}

	_transform(chunk: Buffer, encoding: BufferEncoding, callback: TransformCallback) {
		for (const char of chunk.toString()) {
			// 0x0a is \n
			if (char === "\n") {
				this.push(this.lineCharacters.join(""));
				this.lineCharacters = [];
				continue;
			}
			this.lineCharacters.push(char.toString());
		}
		callback();
	}

	_flush(callback: TransformCallback) {
		callback(null, this.lineCharacters.join(""));
	}
}
