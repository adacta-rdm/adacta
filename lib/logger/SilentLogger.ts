import { PassThrough } from "stream";

import { LOG_LEVEL, Logger } from "./Logger";

/* eslint-disable class-methods-use-this */

/**
 * A logger that does not log anything.
 */
export class SilentLogger extends Logger {
	constructor() {
		super({ level: LOG_LEVEL.SILENT, stream: new PassThrough() });
	}

	trace(): void {}

	debug(): void {}

	info(): void {}

	error(): void {}

	warn(): void {}
}
