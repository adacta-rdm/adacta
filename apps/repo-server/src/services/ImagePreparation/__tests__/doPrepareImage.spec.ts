import { createHash } from "node:crypto";
import { Readable } from "stream";

import probe from "probe-image-size";
import { describe, expect, test } from "vitest";

import { doPrepareImage } from "~/apps/repo-server/src/services/ImagePreparation/doPrepareImage";
import { SilentLogger } from "~/lib/logger/SilentLogger";

const TEST_BASE = "https://s3.storage.omegadot.software/test-images";
const TEST_FILES = {
	JPG: `${TEST_BASE}/test.jpg`,
	// The image contains the same data as test.jpg but the EXIF orientation is set to 5
	// (5 = rotate 90Â° CW + flip horizontally)
	JPG_EXIF_ROTATE: `${TEST_BASE}/test_exif_rotate.jpg`,
	PNG: `${TEST_BASE}/test.png`,
	BMP: `${TEST_BASE}/test_300x200.bmp`, // BMP images are quite large, so this one is already smaller
	HEIC: `${TEST_BASE}/test.heic`,
} as const;

const logger = new SilentLogger();

describe("ImagePreparation", () => {
	describe("Supports Image Types + Resizes", () => {
		test.each(Object.keys(TEST_FILES).filter((k) => k != "JPG_EXIF_ROTATE"))(
			"Supports %s",
			async (type) => {
				const url = TEST_FILES[type as keyof typeof TEST_FILES];
				const buffer = await doPrepareImage(
					url,
					`image/${type}`,
					{ type: "jpg", maxDimensions: { width: 100, height: 100 } },
					logger
				);

				// Basic sanity check
				expect(await probeBuffer(buffer)).toStrictEqual({
					width: 100,
					height: 67,
					type: "jpg",
					mime: "image/jpeg",
					wUnits: "px",
					hUnits: "px",
				});

				// Check final result
				expect(md5(buffer)).toMatchSnapshot();
			}
		);
	});

	describe("Supports image where orientation is provided by EXIF data", () => {
		test("Supports JPG with EXIF orientation", async () => {
			const buffer = await doPrepareImage(
				TEST_FILES.JPG_EXIF_ROTATE,
				"image/jpeg",
				{ type: "jpg", maxDimensions: { width: 100, height: 100 } },
				logger
			);

			// Basic sanity check
			expect(await probeBuffer(buffer)).toStrictEqual({
				// NOTE: The width and height are swapped (compared to the other images)
				// because of the EXIF orientation
				width: 67,
				height: 100,
				type: "jpg",
				mime: "image/jpeg",
				wUnits: "px",
				hUnits: "px",
			});

			// Check final result
			expect(md5(buffer)).toMatchSnapshot();
		});
	});
});

function probeBuffer(buf: Buffer) {
	return probe(Readable.from(buf));
}

function md5(content: Buffer) {
	return createHash("md5").update(content).digest("hex");
}
