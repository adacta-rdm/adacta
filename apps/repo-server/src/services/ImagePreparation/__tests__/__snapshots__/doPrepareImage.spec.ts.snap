// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ImagePreparation > Supports Image Types + Resizes > Supports BMP 1`] = `"a071e251c1ee7c57240cb7e5cb5b0371"`;

exports[`ImagePreparation > Supports Image Types + Resizes > Supports HEIC 1`] = `"67826eab30113f9034069c216f7bef50"`;

exports[`ImagePreparation > Supports Image Types + Resizes > Supports JPG 1`] = `"61738aad55db3b1daa0b4b1cbc27e4ac"`;

exports[`ImagePreparation > Supports Image Types + Resizes > Supports PNG 1`] = `"780c04fdb159dd674ff34c2fdf61b933"`;

exports[`ImagePreparation > Supports image where orientation is provided by EXIF data > Supports JPG with EXIF orientation 1`] = `"d098197b298a7dc0e017a2d973c057fd"`;
